modules/managers/galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos3270.manager/src/main/java/dev/galasa/zos3270/spi/Terminal.java

```diff
@@ -155,16 +181,36 @@ public void disconnect() throws TerminalInterruptedException {
             network.close();
             logger.trace("Network closed OK");
         }
-        if (networkThread != null) {
+
+        logger.trace("connect: Thread "+Thread.currentThread().hashCode()+" is about to request the lock for the network thread");
+        synchronized(networkThreadLock) {
+            logger.trace("connect: Thread "+Thread.currentThread().hashCode()+" has acquired the lock for the network thread");
             try {
-                logger.trace("Waiting for network thread to end...");
-                networkThread.join();
-                logger.trace("Network thread ended OK");
-            } catch (InterruptedException e) {
-                logger.warn("Interrupted while waiting for network thread to end. Network thread might still be active.");
-                throw new TerminalInterruptedException("Join of the network thread was interrupted",e);
+                networkThread = new NetworkThread(this, screen, network, network.getInputStream(), this.deviceTypes);
+                networkThread.start();
+            } finally {
+                logger.trace("connect: Thread "+Thread.currentThread().hashCode()+" is about to release the lock for the network thread");
+            }
+        }
+
+        logger.trace("disconnect: Thread "+Thread.currentThread().hashCode()+" is about to request the lock for the network thread");
+        synchronized(networkThreadLock) {
+            logger.trace("disconnect: Thread "+Thread.currentThread().hashCode()+" has acquired the lock for the network thread");
+            try {
+                if (networkThread != null) {
+                    try {
+                        logger.trace("Waiting for network thread to end...");
+                        networkThread.join(MAX_WAIT_MILLISECONDS_FOR_NETWORK_THREAD_TO_DIE);
```

Comment:
Starting to wonder whether the networkThread does something before it exits that also tries to acquire the same lock as this method. If that happens, then both will enter a waiting state.

Once this 'timeout' is reached and the lock is released, the networkThread will continue, could that lead to another race condition?


Probably need to go through the disconnect-related code paths to see what actually happens.

Reply:
Also not sure how we can unit test this reliably - can we spin up a few network threads in a test and ask a terminal to disconnect?

Reply:
Backed-out the locking.